src/queue/memoryQueue.ts

```diff
@@ -0,0 +1,105 @@
+import { QueueId, QueueInterface } from "./queue";
+import { DSNPMessage, DSNPType } from "../messages/messages";
+
+const InvalidId = new Error("The Queue Id provided is malformed.");
+const IdDoesNotExist = new Error("No message matching the given ID exists.");
+
+interface QueueItem {
+  id: number;
+  message: DSNPMessage;
+}
+
+/**
+ * MemoryQueue provides a simple, in-memory store for DSNP messages to be
+ * announced. This adapter is the default if no other is provided in
+ * configuration.
+ */
+export default class MemoryQueue implements QueueInterface {
+  queues: Record<string, QueueItem[]>;
+  nextId: number;
+
+  constructor() {
+    this.queues = {};
+    this.nextId = 0;
+  }
+
+  /**
+   * enqueue() implements the enqueue function of the QueueInterface. It takes a
+   * DSNP message, adds it to an in-memory store and returns an id which can be
+   * used to reverse this action, if needed.
+   *
+   * @param message The DSNP message to queue up for the next batch
+   * @returns       A queue id to be used to remove the message if needed
+   */
+  async enqueue(type: DSNPType, message: DSNPMessage): Promise<QueueId> {
+    const typeString = type.toString(16);
+    const queueItem = this._createQueueItem(message);
+    const queueId = `${typeString}:${queueItem.id.toString(16)}`;
+
+    if (this.queues[typeString] === undefined) this.queues[typeString] = [];
+    this.queues[typeString].push(queueItem);
+
+    return queueId;
+  }
+
+  /**
+   * dequeue() implements the dequeue function of the QueueInterface. It takes a
+   * DSNP type, then removes and returns the first DSNP message of the matching
+   * type added to the queue. If no messages of the type exist, null will be
+   * returned to indicate the end of the queue.
+   *
+   * @param type The DSNP message type to dequeue
+   * @returns    The dequeued DSNP message or null to indicate end of queue
+   */
+  async dequeue(type: DSNPType): Promise<DSNPMessage | null> {
```

Comment:
Something you mentioned in standup made me wonder. If dequeue takes a type, how do I know what all the types are to check for?

Reply:
Users could iterate through all the DSNPTypes. If a user passes a type to `dequeue` that isn't in the store, it will return null indicating no messages of that type.

If we wanted to check if something is in the queue without removing it, we could implement `peek()` as @enddynayn [suggested above](https://github.com/LibertyDSNP/sdk-ts/pull/40#discussion_r643464365).